torch/_dynamo/variables/builder.py

```diff
@@ -2211,8 +2225,36 @@ def _automatic_dynamic(
                             e.size(i),
                             dim,
                         )
+                        has_size_changed = True
                         frame_state_entry.size[i] = None
 
+                # We want to trigger automatic dynamism when strides change, but we have to think whether stride should
+                # be INFER_STRIDE or DYNAMIC.
+                #
+                # Case 1: if strides change because of size changes, we might not want to allocate a new symbol for
+                # stride. Lets say we have a tensor (10, 20) and we mark the dim=1 dynamic for size. Resulting size will
+                # be (10, s0) and stride can be either (s0, 1) or (s1, 1). In most cases, (s0, 1) is preferred because
+                # users are not changing both size and stride.
+                #
+                # Case 2: But for another case, lets suppose the size remains same between the two invocations but stride
+                # change. In this case, we definitely want to mark the changing stride to be DYNAMIC.
+
+                # Here, we use a hueristic to simplify determination of dynamic stride. For case 1, we will always
+                # assume that stride will be inferred (INFER_STRIDE). This might be suboptimal, where user is doing something
+                # arbitrary size and stride resizing, and we fail to trigger dynamism, but we have not seen any cases
+                # yet. For case 2, we will mark the changing dimensions DYNAMIC.
```

Comment:
It's not clear to me that this needs to be fixed in this PR, but I think the determination algorithm here would be simplified if instead of modeling it as a incremental update process, we say the algorithm takes a list of tensor size+strides, and then determines what should be static/dynamic.

In this model, I think the algorithm should be something like:

1. Assess if sizes are static/dynamic, by comparing size across all entries
2. Compute the inferred strides for all of the tensors individually. If the ordering is ambiguous, consider all of these as possible. If any stride can't be inferred, that forces that stride to be dynamic across all dimensions.
3. Intersect the possible inferred strides for all tensors in the list. If this intersection is non-empty, consistently pick one order as tiebreak and use that for everything. Otherwise, if there is unreconciliable disagreement, kick out those strides as fully dynamic and then try again. This is an overly complicated description, but the basically the idea is to try to define some sort of lattice of symbol allocations and then pick the principal type

This is kind of irritatingly complicated to do, because the stride inference algorithm currently lives in symbolic_shapes.py, and we'd have to redo all the APIs. So once I understand your heuristic, I might just say that it's OK.

Reply:
So, it sounds like your heuristic is something like, if ALL sizes of the tensor are inferred to be static, then check to see if strides have varied. But if ANY size in the tensor is dynamic, don't bother trying to compute automatic dynamic on stride.

This seems... suspicious. Could you walk me through, step-by-step, how your heuristic computes on the repro test you have here?

Reply:
The heuristic is kind of simple and dumb.

First - what is INFER_STRIDE. The way I have implemented it means - let symbolic_shape infra decide if its actually static (because all size dims are static) or dynamic/duck (reuse size symbols whenever possible, or create a new symbol when tensor is strided in an unusual manner).

So, automatic_dynamic job here is to mark the stride either INFER_STRIDE or DYNAMIC.

1) If all the sizes are static, mark stride INFER_STRIDE. Let symbolic shapes decide what to do here.
2) If any of the size changes, still mark stride INFER_STRIDE. Rationale is - if size is changing, we will mark size dim DYNAMIC and symbolic_shapes will then use then reuse size symbols for the strides. Changing size usually changes strides and in predictable manner (contiguous tensors).
3) If size remains same, but the strides change (this is whats happening in generate). Mark the strides DYNAMIC.


So, lets go to the example

~~~

        @torch.compile(backend=cnt)
        def fn(x, y):
            return x * y

        for i in range(1, 4):
            x = torch.randn(4, i)

            # create a view for i > 1
            if i == 1:
                x1 = x
            else:
                x1 = x[:, 0:1]

            y = torch.randn(4, 1)
            print(x1.shape, y.shape)
            fn(x1, y)
~~~


The shape of the tensor is always same `(4, 1)` but the stride keeps changing because of the indexing operation. In this case we fall in the 3rd category, only marking the stride DYNAMIC.